api.go

```diff
@@ -253,148 +356,141 @@ func (api *API) addResource(prototype jsonapi.MarshalIdentifier, source CRUD, ma
-						err := res.handleDeleteToManyRelation(w, r, ps, relation)
-						if err != nil {
-							handleError(err, w, r, marshalers)
-						}
-					}
+					return res.getMiddleware(func(r Request) response {
+						return  res.handleDeleteToManyRelation(r, relation)
+					})
 				}(relation))
 			}
 		}
 	}
 
-	api.router.POST(api.prefix+name, func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-		err := res.handleCreate(w, r, api.prefix, api.info)
-		if err != nil {
-			handleError(err, w, r, marshalers)
-		}
-	})
+	api.router.POST(api.prefix+name, res.getMiddleware(func(r Request) response {
+		return res.handleCreate(r)
+	}))
 
-	api.router.DELETE(api.prefix+name+"/:id", func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-		err := res.handleDelete(w, r, ps)
-		if err != nil {
-			handleError(err, w, r, marshalers)
-		}
-	})
+	api.router.DELETE(api.prefix+name+"/:id", res.getMiddleware(func(r Request) response {
+		return res.handleDelete(r)
+	}))
 
-	api.router.PATCH(api.prefix+name+"/:id", func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-		err := res.handleUpdate(w, r, ps)
-		if err != nil {
-			handleError(err, w, r, marshalers)
-		}
-	})
+	api.router.PATCH(api.prefix+name+"/:id", res.getMiddleware(func(r Request) response {
+		return res.handleUpdate(r)
+	}))
 
 	api.resources = append(api.resources, res)
 
 	return &res
 }
 
-func buildRequest(r *http.Request) Request {
-	req := Request{PlainRequest: r}
-	params := make(map[string][]string)
-	for key, values := range r.URL.Query() {
-		params[key] = strings.Split(values[0], ",")
+func (res *resource) marshalData(data interface{}, r *http.Request) ([]byte, string, error) {
+	marshaler, contentType := selectContentMarshaler(r, res.marshalers)
+	result, err := marshaler.Marshal(data)
+	if err != nil {
+		return nil, "", err
 	}
-	req.QueryParams = params
-	req.Header = r.Header
-	return req
+
+	return result, contentType, nil
+}
+
+func (res *resource) marshalError(data error, r *http.Request) ([]byte, string, int) {
+	marshaler, contentType := selectContentMarshaler(r, res.marshalers)
+
+	status, result := marshaler.MarshalError(data)
+
+	return result, contentType, status
 }
 
-func (res *resource) handleIndex(w http.ResponseWriter, r *http.Request, info information) error {
+func (res *resource) handleIndex(r Request) response {
 	pagination := newPaginationQueryParams(r)
 	if pagination.isValid() {
 		source, ok := res.source.(PaginatedFindAll)
 		if !ok {
-			return NewHTTPError(nil, "Resource does not implement the PaginatedFindAll interface", http.StatusNotFound)
+			return response{
+				Status: http.StatusNotFound,
+				Error: NewHTTPError(nil, "Resource does not implement the PaginatedFindAll interface", http.StatusNotFound),
+			}
 		}
 
-		count, response, err := source.PaginatedFindAll(buildRequest(r))
+		count, resp, err := source.PaginatedFindAll(r)
 		if err != nil {
-			return err
+			return response{
```

Comment:
this could be avoided by just returning err and let the outer `handleError` call in place